inc/dashboard/namespace.php

```diff
@@ -0,0 +1,436 @@
+		$data[ $block_id ]['avg_conversion_rate'] = calculate_average_conversion_rate( [
+			'views' => $views,
+			'conversions' => $conversions,
+		] );
+	}
+
+	return $data;
+}
+
+/**
+ * Sort data by conversion rate.
+ *
+ * @param array $list The array of analytics data by block id.
+ * @param string $order The order to sort by. Accepted values are 'asc' or 'desc'.
+ *
+ * @return array The sorted array.
+ */
+function sort_by_conversion_rate( array $list, string $order = 'desc' ) : array {
+	// If an invalid value was passed to $order, default to 'desc'.
+	if ( ! in_array( $order, [ 'asc', 'desc' ], true ) ) {
+		$order = 'desc';
+	}
+
+	$avg_conversion_rate = array_column( $list, 'avg_conversion_rate' );
+	$sort_order = 'SORT_' . strtoupper( $order );
+
+	array_multisort( $avg_conversion_rate, $sort_order, $list );
+
+	return $list;
+}
+
+/**
+ * Get data from a single block.
+ *
+ * @uses get_views_List()
+ *
+ * @param string $block_id Get a single block's analytics data pulled from the ES query.
+ *
+ * @return array The block's analytics data.
+ */
+function get_block_data( string $block_id ) : array {
+	$data = get_views_list();
+	return $data[ $block_id ];
+}
+
+/**
+ * Get analytics views list data.
+ *
+ * @todo Integrate date range searches.
+ *
+ * @param string $order How to order the data. Accepted values are 'asc' and 'desc'. Defaults to 'desc'.
+ * @param int $days How many days worth of analytics to fetch. Defaults to 7.
+ */
+function get_views_list( string $order = 'desc', int $days = 7 ) : array {
+	$query = [
+		'query' => [
+			'bool' => [
+				'filter' => [
+					[
+						// Query from the current site.
+						'term' => [
+							'attributes.blogId.keyword' => get_current_blog_id(),
+						],
+					],
+					[
+						// We're just interested in views.
+						'terms' => [
+							'event_type.keyword' => [
+								'experienceView',
+								'conversion',
+							],
+						],
+					],
+				],
+			],
+		],
+		'aggs' => [
+			'blocks' => [
+				'terms' => [
+					// Get the block data. This will give us the key for the block, which is stored as the post slug.
+					'field' => 'attributes.clientId.keyword',
+					'size' => 10000, // Use arbitrary large size that is more than we're likely to need.
+					'order' => [
+						'_count' => $order,
+					],
+				],
+				'aggs' => [
+					'events' => [
+						'terms' => [
+							'field' => 'event_type.keyword',
+						],
+					],
+				],
+			],
+		],
+		'size' => 0,
+	];
+
+	// 1: Sort order, asc or desc.
+	// 2: Start date to query by.
+	// 3: End date to query by.
+	$key = sprintf( 'views:list:%1$s:days:%2$d', $order, $days );
+	$cache = wp_cache_get( $key, 'altis-xbs' );
+
+	if ( $cache ) {
+		return $cache;
+	}
+
+	$result = query( $query );
+
+	// Check for a result before getting aggregated data.
+	$data = ( $result ) ? get_aggregate_data( $result['aggregations']['blocks']['buckets'] ?? [] ) : [];
+
+	// Don't cache anything if we didnt' get a result.
+	if ( ! empty( $data ) ) {
```

Comment:
I learned this recently but getting no data is also a valid response and should be cached. You could check this prior to the `get_aggregate_data()` call. Typically what I've done is to set a lower cache time when there's no data like a minute, then 5 minutes when we do have data.

Reply:
I saw that in the Blocks namespace but didn't (and still don't, really) understand why you'd want to cache nothing for a minute rather than just recheck next time.

Reply:
Performance mainly. It's like asking for data and the service correctly saying "there isn't any" and responding with "I don't believe you". There are cases where there might not be any data and this query gets hammered, especially as each row has callbacks for the column data that use this function. If I'm showing 50 XBs per page and sorting by date or title for example if there are no views then this query will be run 100 times in quick succession, once for each views and conversion rate column.